### PR TITLE
Add Cure

### PR DESCRIPTION
**Pull request by gbalabasquer**

### CODE REVIEW

File: src/cure.sol
Comment (iamchrissmith) [inline src/cure.sol:123]: for consistency with the `addSource` pattern, should we do:

```suggestion
        data[src].amt = _toUint128(SourceLike(src).cure());
        if (data[src].amt > 0) {
             total = _add(total, data[src].amt);   
        } 
```

Comment (gbalabasquer) [reply src/cure.sol:123]: Hmm I think in this case we would be doing two extra calls to SLOAD (which are cheaper now but still more expensive than not doing them). So what we can do is:
```suggestion
    data[src].amt = amt = _toUint128(SourceLike(src).cure());
    if (amt > 0) {
        total = _add(total, amt);   
    }
```

Comment (iamchrissmith) [reply src/cure.sol:123]: oh yeah that works

**Comment by talbaneth** [comment]
I'm wondering if we want to time the Cure.reset() calls to be done only at a specific time in the ES process.
This is imprtant since we don't want to use non updated data from the sources.
Also unless I'm missing something there's no use for this module before the ES process.
For that I think we should assume that the data in the sources will not change after the ES waiting period, so it's fine to read at that point. 

To implement this we can add to End a new function before thaw() to call into Cure and have a flag in Cure that allows reset only after that (this also means we will have to auth End on Cure)
That new function in End will be the one checking the waiting period and blocking thaw() with a new flag in End.

**Comment by gbalabasquer** [comment]
> I'm wondering if we want to time the Cure.reset() calls to be done only at a specific time in the ES process. This is imprtant since we don't want to use non updated data from the sources. Also unless I'm missing something there's no use for this module before the ES process. For that I think we should assume that the data in the sources will not change after the ES waiting period, so it's fine to read at that point.
> 
> To implement this we can add to End a new function before thaw() to call into Cure and have a flag in Cure that allows reset only after that (this also means we will have to auth End on Cure). That new function in End will be the one checking the waiting period and blocking thaw() with a new flag in End.
> 
> at thaw() stage Cure will need to then check that all sources have been updated. Can keep a counter for that, and allow each source to be updated only once (thus simplifying the code a bit).

Well it is certainly another approach which might have some advantages with the trade-off of complicating a bit the module and it is relationship with `End` (which is super simple now).
However a part from those extra complexities we would be introducing again the dependency that if any of the `reset` calls fail due a failure in `module.cure()` the whole thing will be stuck. I know we could put a `try/catch` but starts to feel a bit over-engineered.  Anyway let's see what others think about it.

File: src/cure.sol
Comment (talbaneth) [inline src/cure.sol:92]: Might be controversial but for very rarely called functions (which is all logic in this contract :)) I would remove the gas optimization of checking if X > 0 each time before updating total.
Will make the code more compact, and I think that's also the general approach taken in other core contracts.

**Comment by talbaneth** [comment]
> > I'm wondering if we want to time the Cure.reset() calls to be done only at a specific time in the ES process. This is imprtant since we don't want to use non updated data from the sources. Also unless I'm missing something there's no use for this module before the ES process. For that I think we should assume that the data in the sources will not change after the ES waiting period, so it's fine to read at that point.
> > To implement this we can add to End a new function before thaw() to call into Cure and have a flag in Cure that allows reset only after that (this also means we will have to auth End on Cure). That new function in End will be the one checking the waiting period and blocking thaw() with a new flag in End.
> > at thaw() stage Cure will need to then check that all sources have been updated. Can keep a counter for that, and allow each source to be updated only once (thus simplifying the code a bit).
> 
> Well it is certainly another approach which might have some advantages with the trade-off of complicating a bit the module and it is relationship with `End` (which is super simple now). However a part from those extra complexities we would be introducing again the dependency that if any of the `reset` calls fail due a failure in `module.cure()` the whole thing will be stuck. I know we could put a `try/catch` but starts to feel a bit over-engineered. Anyway let's see what others think about it.

Certainly agree about preventing any chance this will block ES.
As discussed one scenario where timing is critical is making sure we sample/reset only after L2 ES hanshake protocol is completed (where I think the waiting period is assumed to be larger than the L2 slow path, but not sure by how much).

File: src/cure.sol
Comment (talbaneth) [inline src/cure.sol:72]: Neat: consider changing `Length` to `Count`.
(Length tends to refer to contiguous elements while Count tends to refer to a number of elements)

File: src/cure.sol
Comment (talbaneth) [inline src/cure.sol:75]: Consider addint a getter for the list (as in https://github.com/makerdao/dss/blob/master/src/clip.sol#L435).

File: src/cure.sol
Comment (talbaneth) [inline src/cure.sol:34]: To try and avoid confusion maybe add a comment such as:
"// position in sources plus 1, 0 means a source does not exist"

SImilarly to:
https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/structs/EnumerableSet.sol#L43

File: src/cure.sol
Comment (talbaneth) [inline src/cure.sol:35]: mm this is a RAD (at least treated as such in END), so cannot be store as uint128.

Comment (talbaneth) [reply src/cure.sol:35]: To avoid the use of uint128 we can use a separate positions array.
Will be more gas costly but I don't think it matters so much specifically in this contract.

Comment (gbalabasquer) [reply src/cure.sol]: Done.

File: src/cure.sol
Comment (gbalabasquer) [reply src/cure.sol]: Done.

File: src/cure.sol
Comment (gbalabasquer) [reply src/cure.sol]: Done.

File: src/cure.sol
Comment (gbalabasquer) [reply src/cure.sol]: Yes you are right, it is doesn't make much sense here. Done.

File: src/cure.sol
Comment (talbaneth) [inline src/cure.sol:86]: Either way is probably ok, but I would remove the cure sampling here, thus separating the logic of adding  a source and sampling it. Just seems to me a bit cleaner.
If we want to sample (/reset) as part of the spell we can just add that call right after.

**Comment by talbaneth** [comment]
Changes look good.
Added another suggestion and a test request.

File: src/cure.sol
Comment (gbalabasquer) [reply src/cure.sol:86]: Yes you have a good point, however I'd prefer to leave this one open for now to see others' opinions before applying the change.

**Comment by talbaneth** [comment]
Looks good to me pending the discussions of timing reset calls and calling reset upon adding a source.

File: src/cure.sol
Comment (hexonaut) [reply src/cure.sol:86]: Yeah I think I'm leaning more towards leaving it out to separate the logic. `reset()` will only need to be called after shutdown really.

File: src/cure.sol
Comment (hexonaut) [inline src/cure.sol:105]: I'm assuming this is here for ESM, and this will be authed on the ESM module for governance attack scenario. Can you confirm that is the only reason?

**Review by hexonaut — CHANGES_REQUESTED** [review]
Some comments on design as well as one critical bug.

I think a non-zero `cure` test should be added to the `end` tests as well.

File: src/cure.sol
Comment (gbalabasquer) [reply src/cure.sol:110]: Oops great catch Sam, can't believe I missed this one.

File: src/cure.sol
Comment (gbalabasquer) [reply src/cure.sol:29]: Yeah let's keep it custom at least in this case.

File: src/cure.sol
Comment (gbalabasquer) [reply src/cure.sol:105]: Yes that's the idea. The other option is to make it authed in the `End` module and call `cure.cage()` in `end.cage()` directly.

File: src/cure.sol
Comment (gbalabasquer) [reply src/cure.sol:110]: Fixed now. I was so focused on the functions to add or remove a source that missed the most important one.

**Comment by gbalabasquer** [comment]
> Some comments on design as well as one critical bug.
> 
> I think a non-zero `cure` test should be added to the `end` tests as well.

Yes planning to add some cure ones on the `end` tests as soon as we finish modeling it.

**Comment by talbaneth** [comment]
@hexonaut @kmbarry1 and others, would be good to hear opinions about whether we want to time the `Cure.reset()` calls to be done only at a specific time in the ES process (and probably rename `Cure.reset()` to something else). See discussion above.
I guess if we do opt for that then it's a big change and better be done sooner rather than later.

File: src/cure.sol
Comment (kmbarry1) [reply src/cure.sol:110]: Looks like the check was implemented as `pos[src] == 1`, but shouldn't it be `pos[src] > 0`? Otherwise `reset` can only ever be called for the first source.

File: src/cure.sol
Comment (kmbarry1) [reply src/cure.sol:105]: I like the idea of calling it in `End.cage` because 1) the End already has to store its address, and 2) this is a single global module. Then no Keeper / non-malicious governance actor needs to worry about caging the Cure separately.

Comment (kmbarry1) [reply src/cure.sol:105]: Although I see some conflicting comments suggesting that maybe we want the Cure to be callable after shutdown in some scenarios--maybe I will answer this for myself as I progress through the review, but if anyone wants to give me the quick rundown that might save time.

File: src/cure.sol
Comment (kmbarry1) [reply src/cure.sol:86]: I don't have terribly strong feelings here, it's almost just a matter of aesthetics. Might personally lean towards a call to `reset` in `addSource` to avoid duplicating logic but quickly confirming that the added source supports the correct API. But really I think any option will work.

File: src/cure.sol
Comment (kmbarry1) [reply src/cure.sol:29]: +1 to @gbalabasquer here, let's not profane `dss` with OZ code (or any other external dependency for that matter).

File: src/end.sol
Comment (kmbarry1) [inline src/end.sol:421]: Maybe a good defensive change to the logic would be:
```
debt = vat.debt() >= cure.total() ? sub(vat.debt(), cure.total()) : vat.debt();
```
?

I am a little conflicted on it b/c while it protects us against certain accidents (e.g. setting the `cure` for some source to an anomalously huge amount), there are still accidents that will break things, using `vat.debt()` as the fallback is unlikely to be correct, and it doesn't really protect from malicious actors (they can still e.g. make `total == debt`). So after going to the trouble of suggesting it maybe I've talked myself out of it already, but going to post anyway for others to consider (maybe it inspires some more useful safeguard e.g. in the Cure itself).

File: src/test/cure.t.sol
Comment (kmbarry1) [inline src/test/cure.t.sol:44]:
```suggestion
```

File: src/cure.sol
Comment (kmbarry1) [reply src/cure.sol:105]: Ah I see we intend to allow `reset` only after `cage`. I believe I am still in favor of my earliest comment here advocating to `cage` it in the End.

**Review by kmbarry1 — CHANGES_REQUESTED** [review]
First pass complete; see comments. At least one small thing to fix (in `reset`).

**Comment by kmbarry1** [comment]
> @hexonaut @kmbarry1 and others, would be good to hear opinions about whether we want to time the `Cure.reset()` calls to be done only at a specific time in the ES process (and probably rename `Cure.reset()` to something else). See discussion above. I guess if we do opt for that then it's a big change and better be done sooner rather than later.

Hmmmm I can see a point to both sides. I think that if we require sources to report during settlement, then we must wrap the calls to `cure()` in try-catch statements so a reverting source cannot block ES. As to whether we _should_ require a post-shutdown report from each source... If `reset` can only be called once per source after shutdown, someone might call it too early for one or more sources, resulting in inaccuracies. So that leads me to think the logic should be "`reset` must be called _at least_ once per source after shutdown" to allow for multiple calls if necessary. Tentatively I think that would be okay (if paired with reversion protection), and I think would make it more likely that the ultimate accounting was correct. Note that a simpler way to implement this (i.e. minimizing changes to the End) is to do it entirely in DssCure by making `total` a `private` variable, and making the public fetching function revert after `cage` unless `reset` has been called at least once for all sources (e.g. by keeping a counter similar to Tal's original suggestion, which is only updated by `reset` post-`cage`). What does everyone else think?

**Comment by kmbarry1** [comment]
Maybe the biggest business logic question here is: _should_ it be possible for a single source to block settlement until it achieves some consistency point? I tend to favor "no" (and this is consistent with how the End relies only on a waiting period at the moment for auctions to finish), but if we think the answer is actually "yes", there are a number of compromises and implementation options we can work through (e.g. make it configurable which sources are strictly required to successfully return a value post-`cage`).

**Comment by talbaneth** [comment]
> > @hexonaut @kmbarry1 and others, would be good to hear opinions about whether we want to time the `Cure.reset()` calls to be done only at a specific time in the ES process (and probably rename `Cure.reset()` to something else). See discussion above. I guess if we do opt for that then it's a big change and better be done sooner rather than later.
> 
> Hmmmm I can see a point to both sides. I think that if we require sources to report during settlement, then we must wrap the calls to `cure()` in try-catch statements so a reverting source cannot block ES. As to whether we _should_ require a post-shutdown report from each source... If `reset` can only be called once per source after shutdown, someone might call it too early for one or more sources, resulting in inaccuracies. So that leads me to think the logic should be "`reset` must be called _at least_ once per source after shutdown" to allow for multiple calls if necessary. Tentatively I think that would be okay (if paired with reversion protection), and I think would make it more likely that the ultimate accounting was correct. Note that a simpler way to implement this (i.e. minimizing changes to the End) is to do it entirely in DssCure by making `total` a `private` variable, and making the public fetching function revert after `cage` unless `reset` has been called at least once for all sources (e.g. by keeping a counter similar to Tal's original suggestion, which is only updated by `reset` post-`cage`). What does everyone else think?

I like this (making `total` private).
Note thoug that the requirement for each source to report after `cage()` might not be strict enough since some of the modules will only start their reporting/calculating process after they are caged. We might want to require each source to report after `End`'s waiting period. That can be achieved either by storing `End` in `Cure` and reading `End.wait`, or even simpler by just passing `End.wait` as a parameter to `Cure.cage()`. However we obtain the waiting duration in `Cure`, it will start it's reporting counter only after the time has passed. If we do it this way then since both `Cure.reset` and `End.thaw` are permissionless there is really no point in allowing each source to be reset more than once.

File: src/cure.sol
Comment (hexonaut) [reply src/cure.sol:29]: Just to be clear I suggest copying the code instead of adding some external dependancy such as [here](https://github.com/makerdao/dss-wormhole/blob/master/src/utils/EnumerableSet.sol). In this case I can appreciate the argument against. I'm fine with custom code. Will review it carefully.

File: src/cure.sol
Comment (gbalabasquer) [reply src/cure.sol:110]: Seriously considering taking vacation... (:man_facepalming:). Yes the check should be the same one than in `delSource` and now is fixed.

**Comment by gbalabasquer** [comment]
Hmmm I don't know, I feel we should not over-engineer it (at least for a first version).
The `End` module already heavily depends on keepers. If keepers do not skim underwater CDPs the whole accounting will be screwed. I think this should be implemented in a similar way.
But if you guys feel too strongly about making sure that at least `reset` (or maybe it should just be call `set`) is called at least once per source after `cage` then I'll proceed implementing it.

**Comment by kmbarry1** [comment]
> We might want to require each source to report after `End`'s waiting period.

It sounds like we're still uncertain about the requirements here. To deal with that uncertainty, I'd suggest a configurable delay in `DssCure` (as opposed to locking to `End.wait`), unless we're willing to accept needing to upgrade the Cure in the future. Upgrades shouldn't be too painful unless we have a lot of sources, which won't be the case at first, so probably it's fine either way.

**Comment by kmbarry1** [comment]
Copying over some discussion from the closed PR (in order, Kurt then Gonza then Tal):

> > > I'd break the open questions down as follows:
> > > 
> > > 1. Must all sources be successfully set **at least** once after shutdown?
> > > 2. Must all sources be successfully set **at most** once after shutdown?
> > > 3. Should the Cure attempt to prevent a source from being set "too soon" after shutdown, or should that logic live in the source itself?
> > > 4. Is it acceptable for global settlement to be blocked or delayed by a source that has not been set yet?
> > 
> > 
> > 
> > 1. In theory yes, as otherwise the accounting will be probably broken (excepting the missing source correspond to report 0 value). However it is better to give the DAI holders less than corresponds than giving nothing because the module is stuck due a broken source.
> > 2. I don't see the need to have that constraint.
> > 3. I think it can live in each source. The source might revert on `cure()` if not ready for final number.
> > 4. I think I wouldn't do it. GS needs to happen and not be totally stuck as it can happen if a `src` is not returning for any reason. Maybe what it can be done is there is a max timeout where if all the sources didn't report, the module still reports whatever is calculated as `total`.
> 
> Agree with almost everything. I like the max timeout idea as it removes the need for a try-catch.
> 
> Just regarding (3) - we can have that revert logic in the sources but it seems a bit like duplication (each module should have it) and source for bugs (each such source writer should be aware to do it and cannot just rely on its `cure` storage variable being constantly updated). If we have the max timeout than we store the cage timestamp anyway. So adding another delay storage parameter and checking against it in `set` seems like a small price to pay for being able to control everything from `Cure`.

I think the discussions boil down to two options:
1) Go with the current simple implementation and rely on Keepers doing what they need to do and `wait` being set long enough to allow sources to settle.
2) Implement an "at least once" requirement for sources to report post-shutdown along with a separate waiting period after which lack of reporting no longer blocks reading `total`.

(1) is simpler but (2) gives us a bit more flexibility (can let `wait` be shorter than the max reporting time for Cure sources). I kind of think (1) is probably sufficient as the incentives for Keepers to actually take all required actions isn't that much different between the two. Unless we think it's very important to be able to speed shutdown up in the case that all sources report well before their deadline.

**Comment by talbaneth** [comment]
> Copying over some discussion from the closed PR (in order, Kurt then Gonza then Tal):
> 
> > > > I'd break the open questions down as follows:
> > > > 
> > > > 1. Must all sources be successfully set **at least** once after shutdown?
> > > > 2. Must all sources be successfully set **at most** once after shutdown?
> > > > 3. Should the Cure attempt to prevent a source from being set "too soon" after shutdown, or should that logic live in the source itself?
> > > > 4. Is it acceptable for global settlement to be blocked or delayed by a source that has not been set yet?
> > > 
> > > 
> > > 
> > > 1. In theory yes, as otherwise the accounting will be probably broken (excepting the missing source correspond to report 0 value). However it is better to give the DAI holders less than corresponds than giving nothing because the module is stuck due a broken source.
> > > 2. I don't see the need to have that constraint.
> > > 3. I think it can live in each source. The source might revert on `cure()` if not ready for final number.
> > > 4. I think I wouldn't do it. GS needs to happen and not be totally stuck as it can happen if a `src` is not returning for any reason. Maybe what it can be done is there is a max timeout where if all the sources didn't report, the module still reports whatever is calculated as `total`.
> > 
> > 
> > Agree with almost everything. I like the max timeout idea as it removes the need for a try-catch.
> > Just regarding (3) - we can have that revert logic in the sources but it seems a bit like duplication (each module should have it) and source for bugs (each such source writer should be aware to do it and cannot just rely on its `cure` storage variable being constantly updated). If we have the max timeout than we store the cage timestamp anyway. So adding another delay storage parameter and checking against it in `set` seems like a small price to pay for being able to control everything from `Cure`.
> 
> I think the discussions boil down to two options:
> 
> 1. Go with the current simple implementation and rely on Keepers doing what they need to do and `wait` being set long enough to allow sources to settle.
> 2. Implement an "at least once" requirement for sources to report post-shutdown along with a separate waiting period after which lack of reporting no longer blocks reading `total`.
> 
> (1) is simpler but (2) gives us a bit more flexibility (can let `wait` be shorter than the max reporting time for Cure sources). I kind of think (1) is probably sufficient as the incentives for Keepers to actually take all required actions isn't that much different between the two. Unless we think it's very important to be able to speed shutdown up in the case that all sources report well before their deadline.

If you think we can count on keepers to sample the source (without any incentive and in ES chaos) between the point that  L2 finished reporting their shutdown procedure (say 7 days after cage because of an L2's dispute period) and the End's waiting period (say we push that to 9 days after cage instead of the current 3 days) than ok, we can go with the simpler implementation (1).

**Comment by kmbarry1** [comment]
> If you think we can count on keepers to sample the source (without any incentive and in ES chaos) between the point that L2 finished reporting their shutdown procedure (say 7 days after cage because of an L2's dispute period) and the End's waiting period (say we push that to 9 days after cage instead of the current 3 days) than ok, we can go with the simpler implementation (1).

I think about it like this: the main incentive to make sure sources report is for those who plan to redeem DAI, because under-reporting raises the value of `debt` recorded and thus lowers the amount of collateral disbursed per DAI. I don't see why an extra delay strengthens or weakens that incentive by much--if no one wants to make sources report, they're probably don't mind waiting a couple more days (if the extra delay would be much longer, weeks or months, then it probably does incentivize a speed up). If we think there's a high chance Keepers will just be unaware that sources need to be reported, then that's a strong argument for having "at least once" enforcement + an extra delay, just so they don't forget.

Personally I do think a "sources must report at least once, but after some delay the requirement is voided" logic makes sense from a correctness perspective.

**Comment by gbalabasquer** [comment]
@kmbarry1 @talbaneth please let me know what you think about this last [commit](https://github.com/makerdao/dss/pull/247/commits/2b23cd9fd6a2a61178fce445ead4fdad1ea06cf1)

**Comment by talbaneth** [comment]
I think this is good (did only a logical review of the recent change, did not yet go into varaible names and synthax etc..).

Again, personally I would have made it even stricter by storing an additional wait period which only after that you can `load()`.
But not a must I guess and we can probably add that in the future if we want (a benefit of the pull model is that update is simple).

**Comment by gbalabasquer** [comment]
> I think this is good (did only a logical review of the recent change, did not yet go into varaible names and synthax etc..).
> 
> Again, personally I would have made it even stricter by storing an additional wait period which only after that you can `load()`. But not a must I guess and we can probably add that in the future if we want (a benefit of the pull model is that update is simple).

Yeah, let's try to not complicate more. I'll wait for your final review so we can aim to pass it to CS on Wednesday.

**Review by kmbarry1 — DISMISSED** [review]
LGTM

I think this is good for a first version.

File: src/cure.sol
Comment (talbaneth) [inline src/cure.sol:30]: Name is a bit confusing since it's not End.wait, maybe change to `stretch`?

File: src/cure.sol
Comment (talbaneth) [inline src/cure.sol:31]: Maybe change to `loose`? (just to have something with more meaning)

File: src/cure.sol
Comment (talbaneth) [inline src/cure.sol:35]: Rename to lcount to conform with `count()` below and use just one word?

File: src/cure.sol
Comment (talbaneth) [inline src/cure.sol:36]: Rename to `report_`? (see suggested change of amount() to report() below)

File: src/cure.sol
Comment (talbaneth) [inline src/cure.sol:65]: // There is a small discrepancy between this function and `amount()` where you return without a variable.

File: src/cure.sol
Comment (talbaneth) [inline src/cure.sol:72]: Rename to `report()` to remove confusion with amount and amt?

File: src/cure.sol
Comment (talbaneth) [inline src/cure.sol:101]: Emit an event?

File: src/cure.sol
Comment (talbaneth) [inline src/cure.sol:116]: Emit an event?

File: src/cure.sol
Comment (talbaneth) [inline src/cure.sol:135]: Emit an event?

File: src/cure.sol
Comment (talbaneth) [inline src/cure.sol:133]: Remove space before the `++`?

**Review by talbaneth — COMMENTED** [review]
See some naming suggestions (subjective, fill free to ignore).
Also some minor synthax stuff.
Finally wondering whether we need events for the sources stuff.

File: src/cure.sol
Comment (gbalabasquer) [reply src/cure.sol]: Done

File: src/cure.sol
Comment (gbalabasquer) [reply src/cure.sol:141]: Done.

File: src/cure.sol
Comment (gbalabasquer) [reply src/cure.sol]: Done.

File: src/cure.sol
Comment (gbalabasquer) [reply src/cure.sol]: Done.

File: src/cure.sol
Comment (gbalabasquer) [reply src/cure.sol:30]: Well it's `Cure.wait`. It wouldn't be the first time we use the same variable name for different contracts.

File: src/cure.sol
Comment (gbalabasquer) [reply src/cure.sol:31]: Same than `wait`.

File: src/cure.sol
Comment (gbalabasquer) [reply src/cure.sol:35]: Maybe just `loaded`?

Comment (gbalabasquer) [reply src/cure.sol:35]: LOL yeah... Ok let me think if I can think on some better approach otherwise will follow your suggestion.

Comment (gbalabasquer) [reply src/cure.sol]: I've renamed `loadedNum` to `lCount` (as you suggested) and `count` to `tCount` (total count).
If we figure better namings, we can always change it later.

**Review by talbaneth — DISMISSED** [review]
LGTM

**Review by hexonaut — COMMENTED** [review]
Two minor points. Overall looks good.

File: src/cure.sol
Comment (hexonaut) [reply src/cure.sol:36]: `peek` ? I hate doing the naming lol

Comment (naszam) [reply src/cure.sol:36]: what about `buzz`, `snap`, `howl` or `copy` ?
ref. [`report` synonyms](https://www.merriam-webster.com/thesaurus/report)

Comment (gbalabasquer) [reply src/cure.sol]: I've done some renaming and exposing the variable as requested.

**Review by talbaneth — APPROVED** [review]
Recent changes LGTM

**Review by hexonaut — APPROVED** [review]
LGTM

**Pull request merged**
